src/main/java/org/openstreetmap/atlas/geography/boundary/DynamicGridIndexBuilder.java

```diff
@@ -1,40 +1,245 @@
 package org.openstreetmap.atlas.geography.boundary;
 
 import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.stream.IntStream;
 
 import org.openstreetmap.atlas.tags.ISOCountryTag;
-import org.openstreetmap.atlas.utilities.maps.MultiMap;
+import org.openstreetmap.atlas.utilities.threads.Pool;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.vividsolutions.jts.geom.Envelope;
 import com.vividsolutions.jts.geom.Polygon;
 import com.vividsolutions.jts.geom.TopologyException;
 import com.vividsolutions.jts.index.strtree.STRtree;
 
 /**
- * This builder builds a spatial index using R-tree based on JTS. Instead of equally divided grid
- * index, dynamic grid index use a 2d kd-tree like structure to generate cells dynamically. This
- * reduces the number of cells generated and allow us to generate much smaller cells at boundary to
- * improve search performance by reducing false positive hits.
+ * This builder builds a spatial index using {@link STRtree} based on JTS. Instead of equally
+ * divided grid index, dynamic grid index use a 2d kd-tree like structure to generate cells
+ * dynamically. This reduces the number of cells generated and allow us to generate much smaller
+ * cells at boundary to improve search performance by reducing false positive hits.
  *
  * @author Yiqing Jin
+ * @author mkalender
  */
 public class DynamicGridIndexBuilder extends AbstractGridIndexBuilder
 {
+    /**
+     * A class to hold units of work while building grid index.
+     *
+     * @author mkalender
+     */
+    class GridWorkItem
```

Comment:
If those inner classes are not used outside, I would keep them private instead of package-protected, no?

Reply:
I'll make them private.